CanvasItem shader breaks when using EXTRA_MATRIX (badly converted to GLSL)

Godot 3 beta 2
Windows 10 64 bits
nVidia geforce 940M

The following shader code ([which is in part copied from doc](http://docs.godotengine.org/en/latest/learning/features/shading/shading_language.html#id3)) breaks rendering and logs errors in the console:

```glsl
shader_type canvas_item;

varying vec2 world_pos;

void vertex() {
	// That line in particular. If removed, things go back to normal
	world_pos = (EXTRA_MATRIX * (WORLD_MATRIX * vec4(VERTEX, 0.0, 1.0))).xy;
}

void fragment() {
	COLOR = vec4(world_pos, 0, 1);
}
```

Project:
[ShaderBugWithWorldMatrixAndExtraMatrix.zip](https://github.com/godotengine/godot/files/1610001/ShaderBugWithWorldMatrixAndExtraMatrix.zip)

Log:
```
1: #version 330
2: #define GLES_OVER_GL
3: #define USE_TEXTURE_RECT
4: #define COLOR_USED
5: precision highp float;
6: precision highp int;
7:
8:
9: layout(location=0) in highp vec2 vertex;
10: layout(location=3) in vec4 color_attrib;
11:
12: #ifdef USE_TEXTURE_RECT
13:
14: uniform vec4 dst_rect;
15: uniform vec4 src_rect;
16:
17: #else
18:
19: #ifdef USE_INSTANCING
20:
21: layout(location=8) in highp vec4 instance_xform0;
22: layout(location=9) in highp vec4 instance_xform1;
23: layout(location=10) in highp vec4 instance_xform2;
24: layout(location=11) in lowp vec4 instance_color;
25:
26: #ifdef USE_INSTANCE_CUSTOM
27: layout(location=12) in highp vec4 instance_custom_data;
28: #endif
29:
30: #endif
31:
32: layout(location=4) in highp vec2 uv_attrib;
33:
34: //skeletn
35: #endif
36:
37: uniform highp vec2 color_texpixel_size;
38:
39:
40: layout(std140) uniform CanvasItemData { //ubo:0
41:
42:     highp mat4 projection_matrix;
43:     highp float time;
44: };
45:
46: uniform highp mat4 modelview_matrix;
47: uniform highp mat4 extra_matrix;
48:
49:
50: out mediump vec2 uv_interp;
51: out mediump vec4 color_interp;
52:
53: #ifdef USE_NINEPATCH
54:
55: out highp vec2 pixel_size_interp;
56: #endif
57:
58:
59: #ifdef USE_LIGHTING
60:
61: layout(std140) uniform LightData { //ubo:1
62:
63:     //light matrices
64:     highp mat4 light_matrix;
65:     highp mat4 light_local_matrix;
66:     highp mat4 shadow_matrix;
67:     highp vec4 light_color;
68:     highp vec4 light_shadow_color;
69:     highp vec2 light_pos;
70:     highp float shadowpixel_size;
71:     highp float shadow_gradient;
72:     highp float light_height;
73:     highp float light_outside_alpha;
74:     highp float shadow_distance_mult;
75: };
76:
77:
78: out vec4 light_uv_interp;
79:
80:
81: out vec4 local_rot;
82:
83:
84: #ifdef USE_SHADOWS
85: out highp vec2 pos;
86: #endif
87:
88: const bool at_light_pass = true;
89: #else
90: const bool at_light_pass = false;
91: #endif
92:
93: #ifdef USE_PARTICLES
94: uniform int h_frames;
95: uniform int v_frames;
96: #endif
97:
98:
99: #if defined(USE_MATERIAL)
100:
101: layout(std140) uniform UniformData { //ubo:2
102:
103:
104: };
105:
106: #endif
107: smooth out vec2 m_world_pos;
108:
109:
110: void main() {
111:
112:    vec4 color = color_attrib;
113:
114: #ifdef USE_INSTANCING
115:    mat4 extra_matrix2 = extra_matrix * transpose(mat4(instance_xform0,instance_xform1,instance_xform2,vec4(0.0,0.0,0.0,1.0)));
116:    color*=instance_color;
117:    vec4 instance_custom = instance_custom_data;
118:
119: #else
120:    mat4 extra_matrix2 = extra_matrix;
121:    vec4 instance_custom = vec4(0.0);
122: #endif
123:
124: #ifdef USE_TEXTURE_RECT
125:
126:    if (dst_rect.z < 0.0) { // Transpose is encoded as negative dst_rect.z
127:            uv_interp = src_rect.xy + abs(src_rect.zw) * vertex.yx;
128:    } else {
129:            uv_interp = src_rect.xy + abs(src_rect.zw) * vertex;
130:    }
131:    highp vec4 outvec = vec4(dst_rect.xy + abs(dst_rect.zw) * mix(vertex,vec2(1.0,1.0)-vertex,lessThan(src_rect.zw,vec2(0.0,0.0))),0.0,1.0);
132:
133: #else
134:    uv_interp = uv_attrib;
135:    highp vec4 outvec = vec4(vertex,0.0,1.0);
136: #endif
137:
138:
139: #ifdef USE_PARTICLES
140:    //scale by texture size
141:    outvec.xy/=color_texpixel_size;
142:
143:    //compute h and v frames and adjust UV interp for animation
144:    int total_frames = h_frames * v_frames;
145:    int frame = min(int(float(total_frames) *instance_custom.z),total_frames-1);
146:    float frame_w = 1.0/float(h_frames);
147:    float frame_h = 1.0/float(v_frames);
148:    uv_interp.x = uv_interp.x * frame_w + frame_w * float(frame % h_frames);
149:    uv_interp.y = uv_interp.y * frame_h + frame_h * float(frame / h_frames);
150:
151: #endif
152:
153: #define extra_matrix extra_matrix2
154:
155: {
156:    {
157:            m_world_pos=(*(modelview_matrix*vec4(outvec.xy, 0.0f, 1.0f))).xy;
158:    }
159:
160:
161: }
162:
163:
164: #ifdef USE_NINEPATCH
165:
166:    pixel_size_interp=abs(dst_rect.zw) * vertex;
167: #endif
168:
169: #if !defined(SKIP_TRANSFORM_USED)
170:    outvec = extra_matrix * outvec;
171:    outvec = modelview_matrix * outvec;
172: #endif
173:
174: #undef extra_matrix
175:
176:    color_interp = color;
177:
178: #ifdef USE_PIXEL_SNAP
179:
180:    outvec.xy=floor(outvec+0.5).xy;
181: #endif
182:
183:
184:    gl_Position = projection_matrix * outvec;
185:
186: #ifdef USE_LIGHTING
187:
188:    light_uv_interp.xy = (light_matrix * outvec).xy;
189:    light_uv_interp.zw =(light_local_matrix * outvec).xy;
190: #ifdef USE_SHADOWS
191:    pos=outvec.xy;
192: #endif
193:
194:
195:    local_rot.xy=normalize( (modelview_matrix * ( extra_matrix * vec4(1.0,0.0,0.0,0.0) )).xy  );
196:    local_rot.zw=normalize( (modelview_matrix * ( extra_matrix * vec4(0.0,1.0,0.0,0.0) )).xy  );
197: #ifdef USE_TEXTURE_RECT
198:    local_rot.xy*=sign(src_rect.z);
199:    local_rot.zw*=sign(src_rect.w);
200: #endif
201:
202:
203:
204: #endif
205:
206: }
207:
208:
ERROR: _display_error_with_code: CanvasShaderGLES3: Vertex Program Compilation Failed:
0(157) : error C1216: dereference of non-pointer

   At: drivers\gles3\shader_gles3.cpp:167
ERROR: ShaderGLES3::get_current_version: Method/Function Failed, returning: 0
   At: drivers\gles3\shader_gles3.cpp:353
ERROR: ShaderGLES3::bind: Condition ' !version ' is true. returned: false
   At: drivers\gles3\shader_gles3.cpp:122
ERROR: ShaderGLES3::_get_uniform: Condition ' !version ' is true. returned: -1
   At: .\drivers/gles3/shader_gles3.h:376
ERROR: ShaderGLES3::_get_uniform: Condition ' !version ' is true. returned: -1
   At: .\drivers/gles3/shader_gles3.h:376
ERROR: ShaderGLES3::_get_uniform: Condition ' !version ' is true. returned: -1
   At: .\drivers/gles3/shader_gles3.h:376
ERROR: ShaderGLES3::_get_uniform: Condition ' !version ' is true. returned: -1
   At: .\drivers/gles3/shader_gles3.h:376
ERROR: ShaderGLES3::_get_uniform: Condition ' !version ' is true. returned: -1
   At: .\drivers/gles3/shader_gles3.h:376
ERROR: ShaderGLES3::_get_uniform: Condition ' !version ' is true. returned: -1
   At: .\drivers/gles3/shader_gles3.h:376
ERROR: ShaderGLES3::_get_uniform: Condition ' !version ' is true. returned: -1
   At: .\drivers/gles3/shader_gles3.h:376
ERROR: ShaderGLES3::_get_uniform: Condition ' !version ' is true. returned: -1
   At: .\drivers/gles3/shader_gles3.h:376
ERROR: ShaderGLES3::_get_uniform: Condition ' !version ' is true. returned: -1
   At: .\drivers/gles3/shader_gles3.h:376
```

--- a/drivers/gles3/shader_compiler_gles3.cpp
+++ b/drivers/gles3/shader_compiler_gles3.cpp
@@ -715,7 +715,7 @@ ShaderCompilerGLES3::ShaderCompilerGLES3() {
 
 	actions[VS::SHADER_CANVAS_ITEM].renames["WORLD_MATRIX"] = "modelview_matrix";
 	actions[VS::SHADER_CANVAS_ITEM].renames["PROJECTION_MATRIX"] = "projection_matrix";
-	actions[VS::SHADER_CANVAS_ITEM].renames["EXTRA_MATRIX"] == "extra_matrix";
+	actions[VS::SHADER_CANVAS_ITEM].renames["EXTRA_MATRIX"] = "extra_matrix";
 	actions[VS::SHADER_CANVAS_ITEM].renames["TIME"] = "time";
 	actions[VS::SHADER_CANVAS_ITEM].renames["AT_LIGHT_PASS"] = "at_light_pass";
 	actions[VS::SHADER_CANVAS_ITEM].renames["INSTANCE_CUSTOM"] = "instance_custom";

